src/radical/pilot/pmgr/launching/base.py

```diff
@@ -25,34 +123,1121 @@ def __init__(self, cfg, session):
+        jd_dict.executable            = '/bin/bash'
+        jd_dict.arguments             = shlex.split('-l ./bootstrap_0.sh %s'
+                                                                      % bs_args)
+        jd_dict.working_directory     = pilot_sandbox
+        jd_dict.project               = project
+        jd_dict.output                = 'bootstrap_0.out'
+        jd_dict.error                 = 'bootstrap_0.err'
+        jd_dict.total_cpu_count       = total_cpu_count
+        jd_dict.total_gpu_count       = total_gpu_count
+        jd_dict.total_physical_memory = requested_memory
+        jd_dict.processes_per_host    = cores_per_node
+        jd_dict.spmd_variation        = spmd_variation
+        jd_dict.wall_time_limit       = runtime
+        jd_dict.queue                 = queue
+        jd_dict.candidate_hosts       = candidate_hosts
+        jd_dict.environment           = dict()
+        jd_dict.system_architecture   = dict(system_architecture)
+
+        # job description environment variable(s) setup
+        if self._prof.enabled:
+            jd_dict.environment['RADICAL_PROFILE'] = 'TRUE'
+
+        jd_dict.environment['RADICAL_BASE'] = resource_sandbox
+
+        # for condor backends and the like which do not have shared FSs, we add
+        # additional staging directives so that the backend system binds the
+        # files from the session and pilot sandboxes to the pilot job.
+        jd_dict.file_transfer = list()
+        if not shared_filesystem:
+
+            jd_dict.file_transfer.extend([
+                'site:%s/%s > %s' % (pilot_sandbox, agent_cfg_name,
+                                                    agent_cfg_name),
+                'site:%s/%s.log.tgz > %s.log.tgz' % (pilot_sandbox, pid, pid),
+                'site:%s/%s.log.tgz < %s.log.tgz' % (pilot_sandbox, pid, pid)
+            ])
+
+            if self._prof.enabled:
+                jd_dict.file_transfer.extend([
+                    'site:%s/%s.prof.tgz > %s.prof.tgz' % (pilot_sandbox, pid, pid),
+                    'site:%s/%s.prof.tgz < %s.prof.tgz' % (pilot_sandbox, pid, pid)
+                ])
+
+            for sdist in sdist_names:
+                jd_dict.file_transfer.extend([
+                    'site:%s/%s > %s' % (session_sandbox, sdist, sdist)
+                ])
+
+
+        self._log.debug("Bootstrap command line: %s %s", jd_dict.executable,
+                jd_dict.arguments)
+
+        pilot['jd_dict'] = jd_dict
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _stage_in(self, pilot, sds):
+        '''
+        Run some input staging directives.
+        '''
+
+        resource_sandbox = self._session._get_resource_sandbox(pilot)
+      # session_sandbox  = self._session._get_session_sandbox (pilot)
+        pilot_sandbox    = self._session._get_pilot_sandbox   (pilot)
+        client_sandbox   = self._session._get_client_sandbox()
```

Comment:
these values are already defined within pilot
https://github.com/radical-cybertools/radical.pilot/blob/416b0ff39c0aa1559cfb72b383d9fbc07464cab7/src/radical/pilot/pmgr/launching/base.py#L696-L705

Reply:
Fair, changed it.  Although the `_get_*_sandbox` methods also cache, so it's no overhead really.